node/actors/network/src/gossip/mod.rs

```diff
@@ -103,4 +117,72 @@ impl Network {
         })
         .await;
     }
+
+    /// Task that keeps hearing about new votes and updates the L1 batch qc.
+    /// It will propagate the QC if there's enough votes.
+    pub(crate) async fn update_batch_qc(&self, ctx: &ctx::Ctx) -> anyhow::Result<()> {
+        // TODO This is not a good way to do this, we shouldn't be verifying the QC every time
+        // Can we get only the latest votes?
+        let attesters = self.genesis().attesters.as_ref().context("attesters")?;
+        loop {
+            let mut sub = self.batch_votes.subscribe();
+            let votes = sync::changed(ctx, &mut sub)
+                .await
+                .context("batch votes")?
+                .clone();
+
+            // Check next QC to collect votes for.
+            let new_qc = self
+                .last_viewed_qc
+                .clone()
+                .map(|qc| {
+                    attester::BatchQC::new(
+                        attester::Batch {
+                            number: qc.message.number.next(),
+                        },
+                        self.genesis(),
+                    )
+                })
+                .unwrap_or_else(|| {
+                    attester::BatchQC::new(
+                        attester::Batch {
+                            number: attester::BatchNumber(0),
+                        },
+                        self.genesis(),
+                    )
+                })
+                .context("new qc")?;
+
+            // Check votes for the correct QC.
+            for (_, sig) in votes.0 {
+                if self
+                    .batch_qc
+                    .clone()
+                    .entry(new_qc.message.number.clone())
+                    .or_insert_with(|| {
+                        attester::BatchQC::new(new_qc.message.clone(), self.genesis()).expect("qc")
+                    })
+                    .add(&sig, self.genesis())
+                    .is_err()
+                {
+                    // TODO: Should we ban the peer somehow?
+                    continue;
+                }
+            }
+
+            let weight = attesters.weight(
+                &self
+                    .batch_qc
+                    .get(&new_qc.message.number)
+                    .context("last qc")?
+                    .signers,
+            );
+
+            if weight < attesters.threshold() {
+                return Ok(());
+            };
+
+            // If we have enough weight, we can update the last viewed QC and propagate it.
```

Comment:
I don't understand why we return if the weight is _less_ than the threshold. Shouldn't it wait for more votes?

Reply:
I also don't really understand the why the network actor [finishes the update task](https://github.com/matter-labs/era-consensus/pull/76/files#diff-7c22d1cc31d4af9bc4765751853b7c0161b50ebf892cf808a7553e1e951817e6R125-R129) if we return from here. Is that forever?

Reply:
Ah, I see, so this just arbitrarily quits. I didn't get that from the PR description, so I thought it's worth asking why I so totally don't get it.